### PR TITLE
Added Moodle support

### PR DESCRIPTION
**Pull request by FerdinandSu**
Added Moodle support for plugin [projectestac/moodle-local_oauth](https://github.com/projectestac/moodle-local_oauth).

Details:

1. Added project **AspNet.Security.OAuth.Moodle**
2. Added docs _moodle.md_
3. Updated README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -150,6 +150,7 @@ If a provider you're looking for does not exist, consider making a PR to add one
...
 | MailRu | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.MailRu?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.MailRu/ "Download AspNet.Security.OAuth.MailRu from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.MailRu?includePreReleases=true)](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.MailRu "Download AspNet.Security.OAuth.MailRu from MyGet.org") | [Documentation](https://o2.mail.ru/docs#web "MailRu developer documentation") |
+| Moodle | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.Moodle?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.Moodle/ "Download AspNet.Security.OAuth.Moodle from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.Moodle?includePreReleases=true)](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.Moodle "Download AspNet.Security.OAuth.Moodle from MyGet.org") | [Documentation](Moodle "Moodle developer documentation") |
```
Comment (martincostello) [inline README.md]: You haven't added a URL for the documentation.

File: docs/moodle.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+# Integrating the Moodle Provider
+
+Moodle end using [projectestac/moodle-local_oauth](https://github.com/projectestac/moodle-local_oauth).
```
Comment (martincostello) [inline docs/moodle.md]: > Moodle end using

What does this mean?

File: docs/moodle.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Example
+
+```csharp
+services.AddAuthentication(options => /* Auth configuration */)
+        .AddMoodle(options =>
+        {
+            //use your own address here
```
Comment (martincostello) [inline docs/moodle.md]:
```suggestion
            // Use your own address here
```

File: docs/moodle.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+            options.UseMoodleSite("https://mymoodlesite.com");
+
+            options.ClientId = "my-client-id";
+            options.ClientSecret = "my-client-secret";
+        });
+```
+
+## Required Additional Settings
+
+_None._
```
Comment (martincostello) [inline docs/moodle.md]: Presumably `UseMoodleSite()` is required, but you've not listed it.

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public static readonly string TokenEndpoint = "http://moodledomain.com/local/oauth/token.php";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+        /// </summary>
+        public static readonly string UserInformationEndpoint = "http://moodledomain.com/local/oauth/user_info.php";
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]: There's no point in including these constants if they point to a domain that doesn't exist.

See the Okta provider for an example of a provider that works against user-provided domains:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/4b83b0a361197290d1faa8f14fc0d59bcdeeaad9/src/AspNet.Security.OAuth.Okta/OktaAuthenticationDefaults.cs#L32-L50

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            TokenEndpoint = MoodleAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = MoodleAuthenticationDefaults.UserInformationEndpoint;
+
+            Scope.Add("user_info");
+
+            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, "username");
+
+            ClaimActions.MapCustomJson(ClaimTypes.Name,
+                e =>
+                    e.GetString("lang")?.StartsWith("zh", System.StringComparison.InvariantCultureIgnoreCase) ?? false ?
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]:
```suggestion
                    e.GetString("lang")?.StartsWith("zh", System.StringComparison.OrdinalIgnoreCase) ?? false ?
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                        $"{e.GetString("lastname")}{e.GetString("firstname")}" :
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]: Should there be a space between the two names? I'm not familiar with exactly how names in China should be written natively.

Would be one of the following based on your answer:

```csharp
e.GetString("lastname") + e.GetString("firstname")
```

or:

```csharp
e.GetString("lastname") + " " + e.GetString("firstname")
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+                    $"{e.GetString("firstname")} {e.GetString("lastname")}");
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]:
```suggestion
                    e.GetString("firstname") + " " + e.GetString("lastname"));
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        /// <param name="moodleSite">Url for your moodle site, like 'http://moodledomain.com'</param>
+        /// <returns>Options changed.</returns>
+        public MoodleAuthenticationOptions UseMoodleSite(string moodleSite)
+        {
+            AuthorizationEndpoint =
+                    $"{moodleSite}/local/oauth/login.php";
+            TokenEndpoint = $"{moodleSite}/local/oauth/token.php";
+            UserInformationEndpoint = $"{moodleSite}/local/oauth/user_info.php";
+            return this;
+        }
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]: Use the pattern the Okta provider uses to configure the endpoints for the user-specified domain by adding a `IPostConfigureOptions<MoodleOptions>` implementation based on a domain provided as a property on the options class.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/4b83b0a361197290d1faa8f14fc0d59bcdeeaad9/src/AspNet.Security.OAuth.Okta/OktaPostConfigureOptions.cs#L19-L44

**Comment by FerdinandSu** [comment]
@martincostello Changes applied according to the reviews.

**Comment by FerdinandSu** [comment]
Sorry, I can't understand what happened to the failed test.

File: src/AspNet.Security.OAuth.Moodle/AspNet.Security.OAuth.Moodle.csproj [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+  <PropertyGroup>
+    <TargetFrameworks>net5.0</TargetFrameworks>
+    <Description>ASP.NET Core security middleware enabling Moodle authentication.</Description>
+    <Authors>HIT-ReFreSH</Authors>
+    <PackageTags>aspnetcore;authentication;moodle;oauth;security</PackageTags>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/AspNet.Security.OAuth.Moodle.csproj:14]: Please restore the JetBrains annotations and use them, otherwise this provider is inconsistent with the others.

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationConstants.cs [hunk 11]
```diff
@@ -0,0 +1,23 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+namespace AspNet.Security.OAuth.Moodle
+{
+    /// <summary>
+    ///     Contains constants specific to the <see cref="MoodleAuthenticationHandler" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationConstants.cs]: Why did you change the indentation of all the XML comments like this to be different to all the code in the rest of the code base?
```suggestion
    /// Contains constants specific to the <see cref="MoodleAuthenticationHandler" />.
```

File: docs/moodle.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+## Example
+
+```csharp
+services.AddAuthentication(options => /* Auth configuration */)
+        .AddMoodle(options =>
+        {
+            options.ClientId = "my-client-id";
+            options.ClientSecret = "my-client-secret";
+            options.Domain = "https://mymoodlesite.com";
```
Comment (martincostello) [inline docs/moodle.md]: This isn't a domain. A domain would be `mymoodlesite.com`.

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationHandler.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(
+            [NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] OAuthTokenResponse tokens)
+        {
+            using var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint);
+            request.Headers.Accept.Add(new ("application/json"));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationHandler.cs]: Please revert these `new` changes because they are then all having to have the space added after them as a workaround for the style analyzer (and then again, makes the style here inconsistent with the rest of the providers).

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationHandler.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+                request,
+                HttpCompletionOption.ResponseHeadersRead,
+                Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                    /* Status: */ response.StatusCode,
+                    /* Headers: */ response.Headers.ToString(),
+                    /* Body: */ await response.Content.ReadAsStringAsync(Context.RequestAborted));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationHandler.cs]:
```suggestion
                                  /* Status: */ response.StatusCode,
                                  /* Headers: */ response.Headers.ToString(),
                                  /* Body: */ await response.Content.ReadAsStringAsync(Context.RequestAborted));
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, "username");
+
+            ClaimActions.MapCustomJson(ClaimTypes.Name,
+                e =>
+                    e.GetString("lang")?.StartsWith("zh", StringComparison.OrdinalIgnoreCase) ?? false
+                        ? e.GetString("lastname") +
+                          e.GetString("firstname")
+                        : e.GetString("firstname") +
+                          " " +
+                          e.GetString("lastname"));
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]:
```suggestion
                        ? e.GetString("lastname") + e.GetString("firstname")
                        : e.GetString("firstname") + " " + e.GetString("lastname"));
```

File: src/AspNet.Security.OAuth.Moodle/MoodlePostConfigureOptions.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+            if (string.IsNullOrWhiteSpace(options.Domain))
+            {
+                throw new ArgumentException("No Moodle domain configured.", nameof(options));
+            }
+
+            options.AuthorizationEndpoint =
+                CreateUrl(options.Domain, MoodleAuthenticationDefaults.AuthorizationEndpointPath);
+            options.TokenEndpoint = CreateUrl(options.Domain, MoodleAuthenticationDefaults.TokenEndpointPath);
+            options.UserInformationEndpoint =
+                CreateUrl(options.Domain, MoodleAuthenticationDefaults.UserInformationEndpointPath);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodlePostConfigureOptions.cs]:
```suggestion
            options.AuthorizationEndpoint = CreateUrl(options.Domain, MoodleAuthenticationDefaults.AuthorizationEndpointPath);
            options.TokenEndpoint = CreateUrl(options.Domain, MoodleAuthenticationDefaults.TokenEndpointPath);
            options.UserInformationEndpoint = CreateUrl(options.Domain, MoodleAuthenticationDefaults.UserInformationEndpointPath);
```

File: src/AspNet.Security.OAuth.Moodle/MoodlePostConfigureOptions.cs [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        private static string CreateUrl(string domain, string path)
+        {
+            // Enforce use of HTTPS
+            var builder = new UriBuilder(domain) { Path = path, Port = -1, Scheme = "https" };
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodlePostConfigureOptions.cs]:
```suggestion
            var builder = new UriBuilder(domain)
            {
                Path = path,
                Port = -1,
                Scheme = "https",
            };
```

File: src/AspNet.Security.OAuth.Moodle/MoodlePostConfigureOptions.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.Extensions.Options;
+
+namespace AspNet.Security.OAuth.Moodle
+{
+    /// <summary>
+    ///     A class used to setup defaults for all <see cref="MoodleAuthenticationOptions" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodlePostConfigureOptions.cs]:
```suggestion
    /// A class used to setup defaults for all <see cref="MoodleAuthenticationOptions" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace AspNet.Security.OAuth.Moodle
+{
+    /// <summary>
+    ///     Defines a set of options used by <see cref="MoodleAuthenticationHandler" />.
+    /// </summary>
+    public class MoodleAuthenticationOptions : OAuthOptions
+    {
+        /// <summary>
+        ///     Initializes a new instance of the <see cref="MoodleAuthenticationOptions" /> class.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]:
```suggestion
        /// Initializes a new instance of the <see cref="MoodleAuthenticationOptions" /> class.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+            ClaimActions.MapJsonKey(Claims.FirstName, "firstname");
+            ClaimActions.MapJsonKey(Claims.LastName, "lastname");
+            ClaimActions.MapJsonKey(Claims.IdNumber, "idnumber");
+            ClaimActions.MapJsonKey(Claims.MoodleId, "id");
+            ClaimActions.MapJsonKey(Claims.Language, "lang");
+        }
+
+        /// <summary>
+        ///     Gets or sets the Moodle domain (Org URL) to use for authentication.
+        ///     For example: 'moodledomain.com'.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]:
```suggestion
        /// Gets or sets the Moodle domain (Org URL) to use for authentication.
        /// For example <c>moodledomain.com</c>.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+using System;
+using System.Security.Claims;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using static AspNet.Security.OAuth.Moodle.MoodleAuthenticationConstants;
+
+namespace AspNet.Security.OAuth.Moodle
+{
+    /// <summary>
+    ///     Defines a set of options used by <see cref="MoodleAuthenticationHandler" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationOptions.cs]:
```suggestion
    /// Defines a set of options used by <see cref="MoodleAuthenticationHandler" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+
+namespace AspNet.Security.OAuth.Moodle
+{
+    /// <summary>
+    ///     Default values used by the Moodle authentication middleware.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
    /// Default values used by the Moodle authentication middleware.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+    /// </summary>
+    public class MoodleAuthenticationDefaults
+    {
+        /// <summary>
+        ///     Default value for <see cref="AuthenticationScheme.Name" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
        /// Default value for <see cref="AuthenticationScheme.Name" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public const string AuthenticationScheme = "Moodle";
+
+        /// <summary>
+        ///     Default value for <see cref="AuthenticationSchemeOptions.ClaimsIssuer" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
        /// Default value for <see cref="AuthenticationSchemeOptions.ClaimsIssuer" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public const string Issuer = "Moodle";
+
+        /// <summary>
+        ///     Default value for <see cref="RemoteAuthenticationOptions.CallbackPath" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
        /// Default value for <see cref="RemoteAuthenticationOptions.CallbackPath" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public const string CallbackPath = "/signin-moodle";
+
+        /// <summary>
+        ///     Default value for <see cref="AuthenticationScheme.DisplayName" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
        /// Default value for <see cref="AuthenticationScheme.DisplayName" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public static readonly string DisplayName = "Moodle";
+
+        /// <summary>
+        ///     Default value for <see cref="OAuthOptions.AuthorizationEndpoint" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public static readonly string AuthorizationEndpointPath = "/local/oauth/login.php";
+
+        /// <summary>
+        ///     Default value for <see cref="OAuthOptions.TokenEndpoint" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
        /// Default value for <see cref="OAuthOptions.TokenEndpoint" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+        /// </summary>
+        public static readonly string TokenEndpointPath = "/local/oauth/token.php";
+
+        /// <summary>
+        ///     Default value for <see cref="OAuthOptions.UserInformationEndpoint" />.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationDefaults.cs]:
```suggestion
        /// Default value for <see cref="OAuthOptions.UserInformationEndpoint" />.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+            [NotNull] string scheme,
+            [NotNull] string caption,
+            [NotNull] Action<MoodleAuthenticationOptions> configuration)
+        {
+            builder.Services
+                .TryAddSingleton<IPostConfigureOptions<MoodleAuthenticationOptions>, MoodlePostConfigureOptions>();
+            return builder.AddOAuth<MoodleAuthenticationOptions, MoodleAuthenticationHandler>(
+                scheme,
+                caption,
+                configuration);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs]:
```suggestion
            builder.Services.TryAddSingleton<IPostConfigureOptions<MoodleAuthenticationOptions>, MoodlePostConfigureOptions>();
            return builder.AddOAuth<MoodleAuthenticationOptions, MoodleAuthenticationHandler>(scheme, caption, configuration);
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+            [NotNull] this AuthenticationBuilder builder,
+            [NotNull] string scheme,
+            [NotNull] Action<MoodleAuthenticationOptions> configuration)
+        {
+            return builder.AddMoodle(scheme, MoodleAuthenticationDefaults.DisplayName, configuration);
+        }
+
+        /// <summary>
+        ///     Adds <see cref="MoodleAuthenticationHandler" /> to the specified
+        ///     <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs]:
```suggestion
        /// Adds <see cref="MoodleAuthenticationHandler" /> to the specified
        /// <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+        public static AuthenticationBuilder AddMoodle(
+            [NotNull] this AuthenticationBuilder builder,
+            [NotNull] Action<MoodleAuthenticationOptions> configuration)
+        {
+            return builder.AddMoodle(MoodleAuthenticationDefaults.AuthenticationScheme, configuration);
+        }
+
+        /// <summary>
+        ///     Adds <see cref="MoodleAuthenticationHandler" /> to the specified
+        ///     <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs]:
```suggestion
        /// Adds <see cref="MoodleAuthenticationHandler" /> to the specified
        /// <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+        /// <param name="builder">The authentication builder.</param>
+        /// <returns>A reference to this instance after the operation has completed.</returns>
+        public static AuthenticationBuilder AddMoodle([NotNull] this AuthenticationBuilder builder)
+        {
+            return builder.AddMoodle(MoodleAuthenticationDefaults.AuthenticationScheme, _ => { });
+        }
+
+        /// <summary>
+        ///     Adds <see cref="MoodleAuthenticationHandler" /> to the specified
+        ///     <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs]:
```suggestion
        /// Adds <see cref="MoodleAuthenticationHandler" /> to the specified
        /// <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    ///     Extension methods to add Moodle authentication capabilities to an HTTP application pipeline.
+    /// </summary>
+    public static class MoodleAuthenticationExtensions
+    {
+        /// <summary>
+        ///     Adds <see cref="MoodleAuthenticationHandler" /> to the specified
+        ///     <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs]:
```suggestion
        /// Adds <see cref="MoodleAuthenticationHandler" /> to the specified
        /// <see cref="AuthenticationBuilder" />, which enables Moodle authentication capabilities.
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs [hunk 35]
```diff
@@ -0,0 +1,85 @@
...
+using System.Diagnostics.CodeAnalysis;
+using AspNet.Security.OAuth.Moodle;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    ///     Extension methods to add Moodle authentication capabilities to an HTTP application pipeline.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs]:
```suggestion
    /// Extension methods to add Moodle authentication capabilities to an HTTP application pipeline.
```

**Review by martincostello — COMMENTED** [review]
Needs unit tests adding.

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationConstants.cs (hunk 11 above)
Comment (FerdinandSu) [reply src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationConstants.cs]: Maybe made by ReSharper....

**Comment by FerdinandSu** [comment]
@martincostello I feel sorry to change the format, which may be caused by my teammate's style analyzer. 

They're **Fixed** now, also I'd added Unit Tests.

But I don't know how to make Tests for special codes for _ClaimTypes.Name_ which will be called only when user-info-Json contains 'lang=zh_XXX'.
I **Tried** to add a _MoodleTestsZh_ class and related data in bundle.json, but Tests in that class failed,
raising the following Exception:

> System.IO.DirectoryNotFoundException : Could not find a part of the path '******\AspNet.Security.OAuth.Providers\artifacts\bin\AspNet.Security.OAuth.Providers.Tests\Debug\net5.0\MoodleZh\bundle.json'.

Could you help me?

File: src/AspNet.Security.OAuth.Moodle/AspNet.Security.OAuth.Moodle.csproj [hunk 36]
```diff
@@ -0,0 +1,18 @@
...
+
+  <PropertyGroup>
+    <Description>ASP.NET Core security middleware enabling Moodle authentication.</Description>
+    <Authors>HIT-ReFreSH</Authors>
+    <PackageTags>aspnetcore;authentication;moodle;oauth;security</PackageTags>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="JetBrains.Annotations" Version="2020.3.0" />
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/AspNet.Security.OAuth.Moodle.csproj]: The version is defined elsewhere, and is for development-time only.

```suggestion
    <PackageReference Include="JetBrains.Annotations" PrivateAssets="All" />
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs [hunk 37]
```diff
@@ -0,0 +1,81 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System;
+using System.Diagnostics.CodeAnalysis;
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs]: Need to swap these out for the `JetBrains.Annotations` versions.

**Comment by martincostello** [comment]
> Could you help me?

If you take a look at the tests for the Apple provider, you'll see that there is one bundle file but some of the configured URLs are changed to make different responses be returned to exercise the different code paths. If you do something similar you should be able to cover the additional code path(s).

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/4b83b0a361197290d1faa8f14fc0d59bcdeeaad9/test/AspNet.Security.OAuth.Providers.Tests/Apple/bundle.json#L35-L47

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/4b83b0a361197290d1faa8f14fc0d59bcdeeaad9/test/AspNet.Security.OAuth.Providers.Tests/Apple/AppleTests.cs#L137

**Comment by FerdinandSu** [comment]
@martincostello `JetBrains.Annotations` versions is ready in _src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationExtensions.cs_.
I also added & passed Tests for special codes now. Thank you for your help.

File: test/AspNet.Security.OAuth.Providers.Tests/Moodle/MoodleTests.cs [hunk 38]
```diff
@@ -0,0 +1,92 @@
...
+        [InlineData(ClaimTypes.GivenName, "三")]
+        [InlineData(ClaimTypes.MobilePhone, "7654321")]
+        [InlineData(Claims.IdNumber, "ZH1234567")]
+        [InlineData(Claims.MoodleId, "22")]
+        [InlineData(Claims.Language, "zh_CN")]
+        public async Task Can_Sign_In_Using_Moodle_In_Chinese(string claimType, string claimValue)
+        {
+            // Arrange
+            static string ChangeToZhUrl(string currentUrl)
+            => currentUrl.Replace("://moodle.local", "://zh.moodle.local", System.StringComparison.InvariantCulture);
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/Moodle/MoodleTests.cs]:
```suggestion
                => currentUrl.Replace("://moodle.local", "://zh.moodle.local", System.StringComparison.InvariantCulture);
```

File: src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationHandler.cs [hunk 39]
```diff
@@ -0,0 +1,63 @@
+﻿/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System.Diagnostics.CodeAnalysis;
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.Moodle/MoodleAuthenticationHandler.cs]: Needs swapping for JetBrains.Annotations.

**Review by martincostello — COMMENTED** [review]
Just two minor changes and then this looks good to merge.

**Comment by FerdinandSu** [comment]
@martincostello The changes required are applied.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for your contribution 🎉 

If you could validate your changes are working as you expect in your project using a version of the provider from [our MyGet feed](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Security.OAuth.Moodle), then if all is OK I'll prepare a 5.0.4 release for NuGet.org.

**Comment by martincostello** [comment]
Version `5.0.4-preview.21178.41` contains the changes from #533.

**Comment by martincostello** [comment]
Hi @FerdinandSukhoi - have you been able to validate the pre-release package?

**Comment by FerdinandSu** [comment]
@martincostello 

@mahoshojoHCG told me that:

> When working with [ASP.NET Core's default AddDefaultIdentity](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/identity?view=aspnetcore-5.0&tabs=visual-studio) and Database, some claims(for example, _NameIdentifier_) will be overwritten by default, and non-latin-names in claim _Name_ **DO NOT** work.

Since most of  the other providers also mapped these claims like mine, I don't think it's a BUG. 
After all, it works quite well on other cases, and the mentioned issue can be fixed by manually _MapJsonKey_ ; But maybe we should warn the users?

**Comment by martincostello** [comment]
Thanks, I will publish the release some time today.

If you want to propose a PR to add something to the documentation about the other thing, feel free.

**Comment by martincostello** [comment]
The Moodle provider is now available from NuGet.org: https://www.nuget.org/packages/AspNet.Security.OAuth.Moodle/5.0.4

Thanks again for your contribution 🎈
